File: python/budgetguard/policy.py
"""The budget envelope for a task. Any limit left as None is not enforced."""

from __future__ import annotations

from dataclasses import dataclass
from typing import Optional


@dataclass(frozen=True)
class BudgetPolicy:
    """Limits applied to a single task.

    max_usd            stop once estimated spend would cross this (needs Pricing)
    max_tokens         total input+output tokens
    max_input_tokens   input tokens only
    max_output_tokens  output tokens only
    max_calls          number of guarded calls
    max_repeats        same call signature this many times -> LoopDetected
    repeat_window      look at the last N signatures when counting repeats
    """

    max_usd: Optional[float] = None
    max_tokens: Optional[int] = None
    max_input_tokens: Optional[int] = None
    max_output_tokens: Optional[int] = None
    max_calls: Optional[int] = None
    max_repeats: Optional[int] = 3
    repeat_window: int = 20

    def __post_init__(self) -> None:
        for name in ("max_usd", "max_tokens", "max_input_tokens", "max_output_tokens", "max_calls", "max_repeats"):
            v = getattr(self, name)
            if v is not None and v <= 0:
                raise ValueError(f"{name} must be positive, got {v!r}")
        if self.repeat_window <= 0:
            raise ValueError("repeat_window must be positive")

    def has_any_limit(self) -> bool:
        return any(
            getattr(self, n) is not None
            for n in ("max_usd", "max_tokens", "max_input_tokens", "max_output_tokens", "max_calls", "max_repeats")
        )

File: typescript/src/policy.ts
/** The budget envelope for a task. Any limit left undefined is not enforced. */

export interface BudgetPolicy {
  /** Stop once estimated spend would cross this (needs Pricing). */
  maxUsd?: number;
  /** Total input+output tokens. */
  maxTokens?: number;
  maxInputTokens?: number;
  maxOutputTokens?: number;
  /** Number of guarded calls. */
  maxCalls?: number;
  /** Same call signature this many times within the window -> LoopDetected. Default 3. */
  maxRepeats?: number;
  /** How many recent signatures to consider when counting repeats. Default 20. */
  repeatWindow?: number;
}

/** Policy with defaults applied. */
export interface ResolvedPolicy extends BudgetPolicy {
  maxRepeats: number | undefined;
  repeatWindow: number;
}

const POSITIVE_KEYS: (keyof BudgetPolicy)[] = [
  "maxUsd",
  "maxTokens",
  "maxInputTokens",
  "maxOutputTokens",
  "maxCalls",
  "maxRepeats",
];

export function resolvePolicy(p: BudgetPolicy = {}): ResolvedPolicy {
  for (const k of POSITIVE_KEYS) {
    const v = p[k];
    if (v != null && (typeof v !== "number" || !(v > 0))) {
      throw new Error(`${k} must be a positive number, got ${JSON.stringify(v)}`);
    }
  }
  const repeatWindow = p.repeatWindow ?? 20;
  if (typeof repeatWindow !== "number" || !(repeatWindow > 0)) {
    throw new Error("repeatWindow must be positive");
  }
  return {
    ...p,
    maxRepeats: "maxRepeats" in p ? p.maxRepeats : 3,
    repeatWindow,
  };
}
